src/cli/virtualfs.ts

```diff
@@ -51,157 +51,106 @@ export function vfsUri(
+  private fileStats = new Map<string, VFSFileStat>();
 
   public knExecutor = new Execute();
 
   // eslint-disable-next-line class-methods-use-this
-  watch(_uri: Uri, _options: { recursive: boolean; excludes: string[] }): Disposable {
-    // ignore, fires for all changes
-    // eslint-disable-next-line @typescript-eslint/no-empty-function
-    return new Disposable(() => {});
+  watch(): Disposable {
+    return new Disposable(() => true);
   }
 
-  async stat(_uri: Uri): Promise<FileStat> {
-    let fileType: FileType = FileType.File;
-    let createTime = 0;
-    let modifiedTime = 0;
-    let fileSize = 0;
+  stat(uri: Uri): FileStat | Thenable<FileStat> {
+    return this.ensureStat(uri);
+  }
 
-    const pathInWorkSpace: string = await getFilePathAsync(this.yamlDirName, _uri.fsPath);
-    fs.stat(pathInWorkSpace, (err, stats) => {
-      if (err) {
-        throw err;
-      }
-      fileType = FileType.File;
-      createTime = stats.ctimeMs;
-      modifiedTime = stats.mtimeMs;
-      fileSize = stats.size;
-    });
+  private ensureStat(uri: Uri): VFSFileStat {
+    if (!this.fileStats.has(uri.toString())) {
+      this.fileStats.set(uri.toString(), new VFSFileStat());
+    }
 
-    return {
-      type: fileType,
-      ctime: createTime,
-      mtime: modifiedTime,
-      size: fileSize,
-    };
-  }
+    const stat = this.fileStats.get(uri.toString());
+    stat.changeStat(stat.size + 1);
 
-  readDirectory(_uri: Uri): [string, FileType][] | Thenable<[string, FileType][]> {
-    return this.readDirectoryAsync();
+    return stat;
   }
 
-  /**
-   * Add the directory if missing.
-   *
-   * Return an array of the array of `[path, fileType]` for the files in the dir.
-   */
-  async readDirectoryAsync(): Promise<[string, FileType][]> {
-    const files: [string, FileType][] = [];
-    await this.createDirectoryAsync(null);
-    const dir = await getFilePathAsync(this.yamlDirName, null);
-    fs.readdirSync(dir).forEach((localFile) => {
-      files.push([path.join(dir, localFile), FileType.File]);
-    });
-    return files;
+  readFile(uri: Uri): Uint8Array | Thenable<Uint8Array> {
+    return this.readFileAsync(uri);
   }
 
-  createDirectory(_uri: Uri): void | Thenable<void> {
-    return this.createDirectoryAsync(_uri);
+  async writeFile(uri: Uri, content: Uint8Array): Promise<void> {
+    const tempPath = os.tmpdir();
+    const fsPath = path.join(tempPath, uri.fsPath);
+    await fsx.ensureFile(fsPath);
+    await fsx.writeFile(fsPath, content);
+    await this.updateK8sResource(fsPath);
+    const exists: boolean = await fsx.pathExists(fsPath);
+    if (exists) {
+      const oldStat = await fsx.stat(fsPath);
+      await fsx.unlink(fsPath);
+
+      // Use timeout to fire file change event in another event loop cycle, this will cause update content inside editor
+      setTimeout(() => {
+        this.fileStats.get(uri.toString())?.changeStat(oldStat.size + 1); // change stat to ensure content update
```

Comment:
What the purpose of `fileStats`? It looks to me that it is never used outside from here. I mean all the things related to `stats`. Missing something or it can be removed? :thinking: